### PR TITLE
Add opt-in multithreading

### PR DESCRIPTION
**Pull request by matthewkeil**
We have a use case for utilizing our level db from multiple worker threads.  The base library did most of the heavy lifting and this PR allows the bindings to be used across workers.

### Methodology

Context-awareness was preserved and so was threadsafe operation from the node.js perspective.  The open and close functionality was all that needed to change and a single mutex protects the setup.  A flag was added so the feature is non-breaking and only provides additional functionality.

The open_db function searches for a record in a `db_handles` map, and if it exists the `open_handle_count` is incremented for that database.  This was used to allow for multiple databases to be accessed per process.  The reverse happens when the destructor or the closed_db method is called.  If the handle is the last to be closed then the db instance is deleted from the map to cleanup the native `LOCK`.

### CODE REVIEW

**Comment by vweevers** [comment]
Interesting! I didn't realize we were this close to supporting this use case.

What's your reasoning for making this opt-in (through `allowMultiThreading`)?

**Comment by matthewkeil** [comment]
> Interesting! I didn't realize we were this close to supporting this use case.
> 
> What's your reasoning for making this opt-in (through `allowMultiThreading`)?

Mostly so the change was non-breaking for existing users @vweevers.  Before I implemented the flag, a few of the unit tests broke so assumed there might be users out there somewhere that are also counting on that behavior.  The flag can be stripped out though or I suppose can default to `true`

**Comment by matthewkeil** [comment]
There is a workflow issue due to python  issue on macos.

https://stackoverflow.com/questions/77233855/why-did-i-got-an-error-modulenotfounderror-no-module-named-distutils

Would you like that fixed @vweevers?

**Comment by vweevers** [comment]
I'm testing a fix: https://github.com/Level/classic-level/commit/fa2a8d7afebf5843dda775e421f7c641c9b53b9e (https://github.com/Level/classic-level/actions/runs/6774425582)

**Comment by vweevers** [comment]
Fixed in main: https://github.com/Level/classic-level/commit/3b1a6f27912b486dcd2a010720db30a4891238ee

**Comment by matthewkeil** [comment]
> Fixed in main: https://github.com/Level/classic-level/commit/3b1a6f27912b486dcd2a010720db30a4891238ee

Would you mind re-running the CI please?

**Comment by vweevers** [comment]
You'll need to rebase your branch to get the fix.

**Comment by vweevers** [comment]
> The flag can be stripped out though

Yeah. We can do that in a future semver-major version. It would simplify the code a bit, but I think it's good to introduce this feature gradually.

File: binding.cc
Comment (vweevers) [inline binding.cc:38]:
```suggestion
// only access this when protected by the handles_mutex!
```

One is enough 😉

File: binding.cc
Comment (vweevers) [inline binding.cc:665]: An inability to close is a bad, inescapable situation, so I prefer not to push that problem to userland. We should either change this to an `assert` (if we're sure it never happens) or just silently return `Status::OK()`.

File: README.md
Comment (vweevers) [inline README.md:204]:
```suggestion
- `allowMultiThreading` (boolean, default: `false`): Allow multiple threads to access the database. This is only relevant when using [worker threads](https://nodejs.org/api/worker_threads.html)
```

No need to qualify it, because we officially only support Node.js atm, and not e.g. bun or Deno.

File: binding.cc
Comment (matthewkeil) [reply binding.cc:38]: Updated

File: index.d.ts
Comment (matthewkeil) [reply index.d.ts:223]: Updated

File: README.md
Comment (matthewkeil) [reply README.md:204]: Updated

File: binding.cc
Comment (matthewkeil) [reply binding.cc]: Updated.  I went with the `OK` route and added a note for those in the future maintainers

**Comment by matthewkeil** [comment]
@vweevers I think I got all of your comments addressed.  Please let me know if there is anything else that you would like updated.

**Comment by vweevers** [comment]
It might be surprising that with `multithreading: false` a handle is still created. E.g. in this example from the tests:

```js
const location = tempy.directory()
const db1 = new ClassicLevel(location)
await db1.open()
t.is(db1.location, location)
```

One would expect to have exclusive access to this db location. But the code that follows shows the opposite:

```js
const db2 = new ClassicLevel(location)
await db2.open({ multithreading: true })
t.is(db2.location, location)
```

In other words, I think `multithreading: false` should also mean "don't allow _others_ to open this db location". Closer to current behavior.

**Comment by matthewkeil** [comment]
> In other words, I think multithreading: false should also mean "don't allow others to open this db location". Closer to current behavior.

Sounds good!! Updated to achieve that behavior @vweevers

File: binding.cc
Comment (vweevers) [inline binding.cc:656]: Error codes are preferred over messages (on the JS side):

> Errors [..] will have a `code` property that is an uppercase string. Error codes will not change between major versions, but error messages will. Messages may also differ between implementations; they are free and encouraged to tune messages.
> — [`abstract-level`](https://github.com/Level/abstract-level#errors)

But I think we can generally simplify the logic here, and fully bypass the handles by default.

What do you think about:

```js
leveldb::Status threadsafe_open(const leveldb::Options &options,
                                bool multithreading,
                                Database &db_instance) {
  // Bypass lock and handles if multithreading is disabled
  if (!multithreading) {
    return leveldb::DB::Open(options, db_instance.location_, &db_instance.db_);
  }

  std::unique_lock<std::mutex> lock(handles_mutex);

  auto it = db_handles.find(db_instance.location_);

  if (it == db_handles.end()) {
    // Database not opened yet for this location, unless it was with
    // multithreading disabled, in which case we're expected to fail here.
    LevelDbHandle handle = {nullptr, 0};
    leveldb::Status status = leveldb::DB::Open(options, db_instance.location_, &handle.db);

    if (status.ok()) {
      handle.open_handle_count++;
      db_instance.db_ = handle.db;
      db_handles[db_instance.location_] = handle;
    }
    
    return status;
  }

  ++(it->second.open_handle_count);
  db_instance.db_ = it->second.db;

  return leveldb::Status::OK();
}

leveldb::Status threadsafe_close(Database &db_instance) {
  std::unique_lock<std::mutex> lock(handles_mutex);
  
  auto it = db_handles.find(db_instance.location_);

  if (it == db_handles.end()) {
    // Was not opened with multithreading enabled
    delete db_instance.db_;
  } else if (--(it->second.open_handle_count) == 0) {
    delete it->second.db;
    db_handles.erase(it);
  }

  // ensure db_ pointer is nullified in Database instance
  db_instance.db_ = NULL;

  return leveldb::Status::OK();
}
```

(I coded this on GitHub so may have mistakes)

Comment (matthewkeil) [reply binding.cc]: Sure.  If that is a better pattern lets go that direction.  Updated

**Review by vweevers — APPROVED** [review]
Nice work! I'll wait with merging for a few days, to give others a chance to review (@juliangruber @ralphtheninja if you're able 🙏).

**Comment by matthewkeil** [comment]
> Nice work! I'll wait with merging for a few days, to give others a chance to review (@juliangruber @ralphtheninja if you're able 🙏).

Hi there @vweevers!  Was curious if @juliangruber or @ralphtheninja were interested in taking a bite at the review apple?  I will be happy to respond to any comments or concerns.

**Comment by ralphtheninja** [comment]
Aaah! So basically we have a layer inbetween db instances and the corresponding location. Looks good!

**Comment by matthewkeil** [comment]
> Aaah! So basically we have a layer inbetween db instances and the corresponding location. Looks good!

Thanks @ralphtheninja!  Much appreciated!!  @juliangruber do you have any comments or concerns?  @vweevers was curious before merging this... Just a nudge so we can get this imported into the [chainsafe/lodestar](https://github.com/ChainSafe/lodestar) repo

**Comment by vweevers** [comment]
That'll do.

**Comment by matthewkeil** [comment]
@vweevers not sure why this unit test failed on the release...

https://github.com/Level/classic-level/actions/runs/6997605304/job/19034756860

**Comment by vweevers** [comment]
Could be a race issue; can you check? I'll wait with the release

**Comment by matthewkeil** [comment]
> Could be a race issue; can you check? I'll wait with the npm release

Yep.  Will take a look this evening

**Pull request merged**
